src/main/scala/scala2021/ashinkarev/task07/Utils.scala

```diff
@@ -0,0 +1,73 @@
+package scala2021.ashinkarev.task07
+
+import scala.util.{Try,Success,Failure}
+import java.io.File
+import java.io.FileInputStream
+
+trait Connection {
+  def close();
+
+  def run();
+}
+
+class PrintConnection(port: Int) extends Connection {
+  def close() = {
+    println("Closed")
+  }
+  def run() = {
+    println("Run")
+  }
+}
+
+object Utils {
+  def withConnection[TResult](port: Int)(body: (Connection) => TResult) = {
+    withResource(() => new PrintConnection(port))(body)((connection) => connection.close())
+  }
+
+  def withFile[TResult](path: String)(body: (File) => TResult) = {
+    withResource(() => new File(path))(body)()
+  }
+
+  def withFileInputStream[TResult](path: String)(body: (FileInputStream) => TResult) = {
+    withResource(() => new FileInputStream(path))(body)((fileInputStream: FileInputStream) => fileInputStream.close())
+  }
+
+  def withResource[TResource, TResult]
+    (resourceFactory: () => TResource)
```

Comment:
It can be written as resourceFactory : => Tresource, When you specify () as input it means there will be no parameters to the function, it is unnecessarily restrict you, it can be ANY function returns Tresult